---
title: "GNXS report"
author: "M. Manzo"
date: '2023-01-17'
output: html_document
params:
  diroi: ''
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r}
library(magrittr)
library(ggplot2)
library(plotly)
library(gnxsQC)
diroi = params$diroi
```


```{r, include=FALSE}
#diroi = '/Users/manzo/Downloads/sgtest0'
sfiles = list.files(path = diroi, recursive = TRUE,
                     full.names = TRUE, pattern = "Snvindel.tsv")

xlsx_files = list.files(path = diroi, recursive = TRUE,
                     full.names = TRUE, pattern = ".*combined_output.xlsx")
xlsx_files = grep("~",xlsx_files, value = TRUE,invert = TRUE)

prep_files = list.files(path = diroi, recursive = TRUE,
                     full.names = TRUE, pattern = "prep_snv.txt")

info_files = list.files(path = diroi, recursive = TRUE,
                     full.names = TRUE, pattern = "Info.csv")
```


```{r}
library(magrittr)
sample_info = dplyr::bind_rows(lapply(info_files, extract_info))
readr::write_tsv(sample_info, paste0(opt$dir,"/GNXS_QC_report/SampleInfoTable.tsv"))

snvs = lapply(sfiles, readin_snvindel)
preps = dplyr::bind_rows(lapply(prep_files, readin_prepfiles))
readr::write_tsv(preps,  paste0(opt$dir,"/GNXS_QC_report/Sample_variants_aggretation.tsv"))

```


## Frequent variants
```{r, eval=TRUE}
prep_insight = preps %>% dplyr::count(variants, sort = TRUE) %>% dplyr::filter(n >1)
DT::datatable(prep_insight)
```


```{r, include=FALSE, eval=FALSE}
snvs = dplyr::bind_rows(snvs)
snvs = snvs %>% tidyr::unite(gene, nucleotide_change, col = "combo_var", sep = "_" ) %>% dplyr::select(-amino_acid_change)

snvs = snvs %>% dplyr::group_by(vartype) %>% dplyr::group_split()
nonsyn = snvs[[1]] %>% tidyr::pivot_wider(names_from = filepath, values_from = allele_frequency) %>%
  dplyr::select(-vartype)

syns = snvs[[2]] %>% tidyr::pivot_wider(names_from = filepath, values_from = allele_frequency) %>%
  dplyr::select(-vartype)

snvs = snvs %>% tidyr::pivot_wider(names_from = filepath, values_from = allele_frequency) %>%
  dplyr::select(-vartype)


```


```{r, include=FALSE, eval=FALSE}
snvs_mat = snvs %>% dplyr::select(-combo_var) %>% as.matrix()
syns_mat = syns %>% dplyr::select(-combo_var) %>% as.matrix()
nonsyn_mat = nonsyn %>% dplyr::select(-combo_var) %>% as.matrix()
rownames(snvs_mat) = snvs$combo_var
rownames(syns_mat) = syns_mat$combo_var
rownames(nonsyn_mat) = nonsyn$combo_var

colnames(snvs_mat) = unname(sapply(colnames(snvs_mat), shorten_name))
colnames(syns_mat) = unname(sapply(colnames(syns_mat), shorten_name))
colnames(nonsyn_mat) = unname(sapply(colnames(nonsyn_mat), shorten_name))
```


```{r, include=FALSE, echo=FALSE, eval=FALSE}
df1 = data.frame(gene = letters[1:10],
                var = 1:10,
                sample = "a")

df2 = data.frame(gene = letters[6:13],
                var = 1:8,
                sample = "b")

dplyr::bind_rows(df2,df1) %>% tidyr::pivot_wider(names_from = sample, values_from = var)

jacc_distance(letters[1:10], letters[6:10])
jacc_distance(names(syns_mat[,1][!is.na(syns_mat[,1])]),
              names(syns_mat[,2][!is.na(syns_mat[,2])]))

```


```{r, fig.height=5, fig.width=5, eval=FALSE}
## Syn -- sample distances

distance_mat = overlap_coef_mat(tdf)
mat_hc = hclust(as.dist(distance_mat), method = "ward.D2")
pheatmap::pheatmap(distance_mat, 
                        col = viridis::viridis(10, option = "D"),
                        cluster_cols = mat_hc,
                        cluster_rows = mat_hc,
                        show_row_names = FALSE)
```

```{r, fig.height=10, fig.width=7, eval=FALSE}
## Syn -- variants

tdf = cbind(snvs_mat, snvs_mat[,1])
tdf[1,9] = NA
tdf[2,9] = NA
distance_mat = overlap_coef_mat(tdf)
mat_hc = hclust(as.dist(distance_mat), method = "ward.D2")

pheatmap::pheatmap(tdf, 
                        col = RColorBrewer::brewer.pal(9, "RdPu"),
                        cluster_cols = mat_hc,
                        cluster_rows = FALSE,
                        show_row_names = TRUE)
```

```{r, fig.height=5, fig.width=5, eval=FALSE}
## Nonsyn -- sample distances

distance_mat = overlap_coef_mat(nonsyn_mat)
mat_hc = hclust(as.dist(distance_mat), method = "ward.D2")
pheatmap::pheatmap(distance_mat, 
                        col = viridis::viridis(10, option = "D"),
                        cluster_cols = mat_hc,
                        cluster_rows = mat_hc,
                        show_row_names = FALSE)


```

```{r, fig.height=10, fig.width=7, eval=FALSE}
## Nonsyn -- variants

pheatmap::pheatmap(nonsyn_mat, 
                        col = RColorBrewer::brewer.pal(9, "RdPu"),
                        cluster_cols = mat_hc,
                        cluster_rows = FALSE,
                        show_row_names = TRUE)
```


```{r}
plot_prep_snvindel = function(filepath){
  snv = readr::read_tsv(filepath)
  snv = janitor::clean_names(snv)
  #snv = dplyr::filter(snv, grepl("PRESENT",call ))
  #snv = dplyr::filter(snv, grepl("PRESENT", call))
  snv = dplyr::filter(snv, call != "ABSENT")
  snv = dplyr::filter(snv, allele_frequency != 0)
  snv = snv %>% dplyr::filter(gene != "CDKN2A-DT") 
  snv = dplyr::mutate(snv, vartype = ifelse(grepl("=", amino_acid_change), "syn","nonsyn"))
  snv = dplyr::select(snv, gene, nucleotide_change,amino_acid_change,phred_qual_score,coverage, allele_frequency)
  snv$filepath = filepath
  snv = snv %>% tidyr::unite(gene,amino_acid_change, nucleotide_change, col = "combo_var", sep = "_" )
  return(snv)
}
plot_prep_table = dplyr::bind_rows(lapply(sfiles, plot_prep_snvindel))

plot_prep_table$combo_var = gsub("_p\\.\\?",'',plot_prep_table$combo_var)

plot_prep_table$filepath = sapply(plot_prep_table$filepath, shorten_name)

af_mat = plot_prep_table %>% dplyr::select(-phred_qual_score, -coverage) %>% tidyr::pivot_wider(names_from = filepath, values_from = allele_frequency )

af.mat = dplyr::select(af_mat, where(is.numeric)) %>% as.matrix()
rownames(af.mat) = af_mat$combo_var
```

## Clustering of samples
```{r}
na_vector = rowMeans(af.mat)
mat_hc = overlap_coef_mat(af.mat[is.na(na_vector),])
mat_hc = hclust(as.dist(mat_hc), method = "ward.D2")
```

## Sample distance matrix
```{r}
pheatmap::pheatmap(overlap_coef_mat(af.mat[is.na(na_vector),]), 
                        col = viridis::viridis(10, option = "D"),
                        cluster_cols = mat_hc,
                        cluster_rows = mat_hc,
                        show_row_names = TRUE)
```

## Allele frequency
```{r, fig.height=10}
pheatmap::pheatmap(af.mat[is.na(na_vector),], 
                        col = RColorBrewer::brewer.pal(9, "RdPu"),
                        cluster_cols = mat_hc,
                        cluster_rows = FALSE,
                        show_row_names = TRUE)

```

## Coverage / Phred score
```{r}
p = plot_prep_table %>% 
  ggplot(aes(phred_qual_score, coverage,group=1, text =combo_var)) + 
  geom_point() + 
  scale_x_log10() + 
  scale_y_log10() + 
  geom_hline(yintercept = 500) + 
  geom_vline(xintercept = 50)

ggplotly(p)
```




